pkg/addon/config.go

```diff
@@ -0,0 +1,214 @@
+	if err != nil {
+		return nil, fmt.Errorf("incorrect es-node-count parameter. %v", err)
+	}
+	mpEsStorage := resource.MustParse(addonConfig.EsStorageSize)
+
+	result := &logging.ClusterLogging{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "instance",
+			Namespace: namespace,
+		},
+		Spec: logging.ClusterLoggingSpec{
+			ManagementState: "Managed",
+			Visualization: &logging.VisualizationSpec{
+				Type: "kibana",
+				KibanaSpec: logging.KibanaSpec{
+					Replicas: 1,
+				},
+			},
+			LogStore: &logging.LogStoreSpec{
+				Type: "elasticsearch",
+				ElasticsearchSpec: logging.ElasticsearchSpec{
+					Resources: &v1.ResourceRequirements{
+						Requests: map[v1.ResourceName]resource.Quantity{
+							"memory": resource.MustParse(addonConfig.EsMemory),
+						},
+					},
+					NodeCount: int32(nodeCount),
+					Storage: elasticsearch.ElasticsearchStorageSpec{
+						StorageClassName: &addonConfig.EsStorageClass,
+						Size:             &mpEsStorage,
+					},
+					RedundancyPolicy: "SingleRedundancy",
+				},
+				RetentionPolicy: &logging.RetentionPoliciesSpec{
+					App: &logging.RetentionPolicySpec{
+						MaxAge: elasticsearch.TimeUnit(addonConfig.EsMaxRetention),
+					},
+					Infra: &logging.RetentionPolicySpec{
+						MaxAge: elasticsearch.TimeUnit(addonConfig.EsMaxRetention),
+					},
+					Audit: &logging.RetentionPolicySpec{
+						MaxAge: elasticsearch.TimeUnit(addonConfig.EsMaxRetention),
+					},
+				},
+			},
+			Collection: &logging.CollectionSpec{
+				Logs: logging.LogCollectionSpec{
+					Type: "fluentd",
+				},
+			},
+			Curation: &logging.CurationSpec{
+				Type: "curator",
+				CuratorSpec: logging.CuratorSpec{
+					Schedule: "30 3 * * *",
+				},
+			},
+		},
+	}
+	return result, nil
+}
+
+func readAddonConfiguration(cl client.Client, namespace, secretName string) (*AddonConfiguration, error) {
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{},
+	}
+	err := cl.Get(context.Background(), client.ObjectKey{
+		Name:      secretName,
+		Namespace: namespace,
+	}, secret)
+
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil, ErrNoAddonConfig
+		} else {
+			return nil, fmt.Errorf("unable to get addon-secret. %v", err)
+		}
+	}
+
+	result := &AddonConfiguration{}
+	sData := secret.Data
+
+	v := reflect.ValueOf(result).Elem()
+	t := reflect.TypeOf(*result)
+	for i := 0; i < v.NumField(); i++ {
+		fieldValue := v.Field(i)
+		fieldType := t.Field(i)
+		tTag := fieldType.Tag
+		if iv, ok := sData[tTag.Get("data-field")]; !ok || len(iv) <= 0 {
+			defaultValue := tTag.Get("default")
+			fieldValue.SetString(defaultValue)
+		} else {
+			fieldValue.SetString(string(iv))
+		}
+	}
+	return result, nil
+}
+
+func newClient() (client.Client, error) {
+	config, err := config.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	shm := scheme.Scheme
+
+	if err := apis.AddToScheme(shm); err != nil {
+		log.Error(err, "failed to add resources to scheme", "resource", "apis")
+		os.Exit(1)
```

Comment:
remove `oc.Exit`

Reply:
+1 - never call os.Exit() unless you're in main() or a trivial helper function called by main. Makes the code un-reusable in any other context and will stop clean-up code from running if any is ever added.

Reply:
Fixed